x-pack/plugins/observability_solution/dataset_quality/server/routes/data_streams/check_and_load_integration/index.ts

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { ElasticsearchClient } from '@kbn/core-elasticsearch-server';
+import { PackageClient } from '@kbn/fleet-plugin/server';
+import { Logger } from '@kbn/logging';
+import { validateCustomComponentTemplate } from './validate_custom_component_template';
+import { getIntegrations } from '../../integrations/get_integrations';
+import { getComponentTemplatePrefixFromIndexTemplate } from '../../../../common/utils/component_template_name';
+import { CheckAndLoadIntegrationResponse } from '../../../../common/api_types';
+import { dataStreamService } from '../../../services';
+
+// The function works on 2 conditions:
+// 1. It checks if integration name is present in meta field response of the datastream.
+// If yes, it considers it to be an integration. No further checks
+// 2. If not, then it does the various checks
+export async function checkAndLoadIntegration({
+  esClient,
+  packageClient,
+  logger,
+  dataStream,
+}: {
+  esClient: ElasticsearchClient;
+  packageClient: PackageClient;
+  logger: Logger;
+  dataStream: string;
+}): Promise<CheckAndLoadIntegrationResponse> {
+  const [dataStreamInfo] = await dataStreamService.getMatchingDataStreams(esClient, dataStream);
+
+  const indexTemplate = dataStreamInfo?.template;
+  const isManaged = dataStreamInfo?._meta?.managed;
+
+  const integrationNameFromDataStream = dataStreamInfo?._meta?.package?.name;
+
+  // Index template must be present and isManaged should be true or
+  // integration name should be present
+  // Else it's not an integration
+  if ((!indexTemplate || !isManaged) && !integrationNameFromDataStream) {
+    return { isIntegration: false, areAssetsAvailable: false };
+  }
+
+  const allIntegrations = await getIntegrations({ packageClient, logger });
```

Comment:
**doubt**: do we have knowledge about how performance expensive is this? If there are many integrations installed, I'm worried it could take longer than expected

Reply:
Are you thinking of trying to fetch the particular integration directly instead? Or do you have a different approach in mind?

Reply:
Exactly, I see that after retrieving all the packages we search for the exact one related to the data stream.
Getting all of them seems more expensive performance-wise, especially given the fact that `getIntegrations` also fetches all the data streams for each integration, making it even more costly.
The `packageClient` used to retrieve the packages also exposes a `getPackage` method to fetch only a targeted package (it's also used inside the dataset retrieval for each integration), it seems a better option to go for since the scale of the installed integrations would not affect this process.

Reply:
Ok, I'll try to make that work :+1: